src/ripple/basics/impl/PerfLogImp.h

```diff
@@ -0,0 +1,222 @@
+#include <ripple/beast/core/SystemStats.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/core/Stoppable.h>
+#include <ripple/protocol/JsonFields.h>
+#include <condition_variable>
+#include <cstdint>
+#include <fstream>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace ripple {
+namespace perf {
+
+/**
+ * Implementation class for PerfLog.
+ */
+class PerfLogImp
+    : public PerfLog, Stoppable
+{
+    /**
+     * Track performance counters and currently executing tasks.
+     */
+    struct Counters
+    {
+    public:
+        /**
+         * RPC performance counters.
+         */
+        struct Rpc
+        {
+            // Keep all items that need to be synchronized in one place
+            // to minimize copy overhead while locked.
+            struct Sync
+            {
+                // Counters for each time a method starts and then either
+                // finishes successfully or with an exception.
+                std::uint64_t started {0};
+                std::uint64_t finished {0};
+                std::uint64_t errored {0};
+                // Cumulative duration of all finished and errored method calls.
+                microseconds duration {0};
+            };
+
+            Sync sync;
+            std::string const label;
+            mutable std::mutex mut;
+
+            Rpc(std::string const& labelArg)
+                : label (labelArg)
+            {}
+
+            Rpc(Rpc const& orig)
+                : sync (orig.sync)
+                , label (orig.label)
+            {}
+        };
+
+        /**
+         * Job Queue task performance counters.
+         */
+        struct Jq
+        {
+            // Keep all items that need to be synchronized in one place
+            // to minimize copy overhead while locked.
+            struct Sync
+            {
+                // Counters for each time a job is enqueued, begins to run,
+                // finishes.
+                std::uint64_t queued {0};
+                std::uint64_t started {0};
+                std::uint64_t finished {0};
+                // Cumulative duration of all jobs' queued and running times.
+                microseconds queuedDuration {0};
+                microseconds runningDuration {0};
+            };
+
+            Sync sync;
+            std::string const label;
+            mutable std::mutex mut;
+
+            Jq(std::string const& labelArg)
+                : label (labelArg)
+            {}
+
+            Jq(Jq const& orig)
+            : sync (orig.sync)
+            , label (orig.label)
+            {}
+        };
+
+        // rpc_ and jq_ do not need mutex protection because all
+        // keys and values are created before more threads are started.
+        std::unordered_map<std::string, Rpc> rpc_;
+        std::unordered_map<std::underlying_type_t<JobType>, Jq> jq_;
```

Comment:
It probably got missed in the back and forth, but underlying_type_t is necessary to let JobType be the key in a hash map, because `int` has a hash function but JobType does not.

Reply:
I checked on my compiler and an `std::unordered_map<JobType, steady_time_point>` compiles just fine.  I also checked cppreference (http://en.cppreference.com/w/cpp/utility/hash).  About half way down it says (since C++14) "... the standard library provides [hash] specializations for all (scoped and unscoped) enumeration types."  So I would expect it to work, since we're using C++14 and `JobType` is an unscoped `enum`.  Is there a particular compiler or tool chain that fails for you?

Reply:
Just to put a cap on this conversation, we found that gcc 5.4 does not implement `hash<>` for `enum`s.  So, in order to use that compiler, we need to leave the `std::underlying_type<>`.